### PR TITLE
Kademlia forwarder peer query service

### PR DESCRIPTION
**Pull request by nolash**

### CODE REVIEW

File: SWIPs/swip-forwarder.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+// also implements context.Context
+type SessionContext struct {
+	CapabilityIndex string
+	SessionId string
+}
+
+type SessionRPCInterface interface {
+	NewForwardPeerSession(ctx context.Context) (context.Context, error)
+	Subscribe(ctx context.Context, "kademlia", ch chan PeerEvent, "forward") (sub rpc.Subscription, err error)
```
Comment (acud) [inline SWIPs/swip-forwarder.md]: can you explain why there are hardcoded values in the signature?

File: SWIPs/swip-forwarder.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	GetForwardPeer(ctx context.Context, numberOfPeers int) (peers []ForwardPeer, err error)
+	CloseForwardPeerSession(ctx context.Context)
```
Comment (acud) [inline SWIPs/swip-forwarder.md]: maybe just CloseSession?

File: SWIPs/swip-forwarder.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	Close()
+}
+
+// also implements context.Context
+type SessionContext struct {
+	CapabilityIndex string
+	SessionId string
+}
+
+type SessionRPCInterface interface {
```
Comment (acud) [inline SWIPs/swip-forwarder.md]: `interface` type name should not end with `Interface`. I'd also vote to reshuffle this name altogether in favor of something more meaningful?

File: SWIPs/swip-forwarder.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+Otherwise, the `Session` terminates immediately after completing the `Get` request.
+
+To initiate a `Session`, `NewForwardPeerSession` should be called with a `SessionContext` as the parameter (a different context will generate an error). The returned context will contain a unique `sessionId` key generated by the server. This context can in turn be passed to `Get` and (optionally) `Subscribe` to attain the corresponding functionality provided by the library interface.
+
+If `NewForwardPeerSession` has been called, `CloseForwardPeerSession` _MUST_ be called with the `SessionContext` updated by the server to terminate the `Session`.
+
+## Rationale
+
+The component shall provide a service where given a specific Swarm Overlay Address, connected peers will be returned in the order of closest to farthest.
```
Comment (acud) [inline SWIPs/swip-forwarder.md]: you mention a `given Swarm Overlay Address`, however a pivot address is not mentioned in any of the function signatures you've defined. In layman terms: you need to know which peer to forward to given a content addressed hash and the peers to forward to would be ordered according to that reference, but that reference parameter is not mentioned in the signatures (or am i getting this all wrong?)

File: SWIPs/swip-forwarder.md (hunk 2 above)
Comment (nolash) [reply SWIPs/swip-forwarder.md]: Is it not more intiutive if it matches the constructor name?

File: SWIPs/swip-forwarder.md (hunk 3 above)
Comment (nolash) [reply SWIPs/swip-forwarder.md]: `ForwardSession` `ForwardSessionAPI` @acud ?
